specification/dartLangSpec.tex

```diff
@@ -11677,113 +12041,209 @@ \subsection{Identifier Reference}
-\item If $d$ is a type parameter $T$, then the value of $e$ is the value of the actual type argument corresponding to $T$ that was passed to the generative constructor that created the current binding of \THIS{}.
-If, however, $e$ occurs inside a static member, a compile-time error occurs.
-
-%\commentary{We are assured that \THIS{} is well defined, because if we were in a static member the reference to $T$ is a compile-time error (\ref{generics}.)}
-%\item If $d$ is a library variable then:
-%  \begin{itemize}
-%  \item If $d$ is of one of the forms \code{\VAR{} $v$ = $e_i$;} , \code{$T$ $v$ = $e_i$;} , \code{\FINAL{} $v$ = $e_i$;} or \code{\FINAL{} $T$ $v$ = $e_i$;} and no value has yet been stored into $v$ then the initializer expression $e_i$ is evaluated. If, during the evaluation of $e_i$, the getter for $v$ is referenced, a \code{CyclicInitializationError} is thrown. If the evaluation succeeded yielding an object $o$, let $r$ be $o$, otherwise let $r$ be the null object (\ref{null}). In any case, $r$ is stored into $v$. The value of $e$ is $r$.
-\item If $d$ is a constant variable of one of the forms
-  \code{\CONST{} $v$ = $e$;} or \code{\CONST{} $T$ $v$ = $e$;}
-  then the value of \id{} is the value of the constant expression $e$.
-%  Otherwise
-%  \item $e$ evaluates to the current binding of \id.
-%  \end{itemize}
-\item If $d$ is a local variable (\commentary{which can be a formal parameter}) then $e$ evaluates to the current binding of \id.
-\item If $d$ is a static method, top-level function or local function then $e$ evaluates to the function object obtained by closurization (\ref{functionClosurization}) of the declaration denoted by $d$.
-\item If $d$ is the declaration of a class variable, static getter or static setter declared in class $C$, then evaluation of $e$ is equivalent to evaluation of the property extraction (\ref{propertyExtraction}) \code{$C$.\id}.
-\item If $d$ is the declaration of a library variable, top-level getter or top-level setter, then evaluation of $e$ is equivalent to evaluation of the top level getter invocation (\ref{topLevelGetterInvocation}) \id.
-\item Otherwise, if $e$ occurs inside a top level or static function (be it function, method, getter, or setter) or variable initializer, evaluation of $e$ causes a \code{NoSuchMethod} to be thrown.
-\item Otherwise, evaluation of $e$ is equivalent to evaluation of the property extraction (\ref{propertyExtraction}) \code{\THIS.\id}.
+\item
+  If $D$ is a class, mixin, type alias or type parameter,
+  the static type of $e$ is \code{Type}.
+\item
+  If $D$ is the declaration of
+  a library variable or getter,
+  the static type of $e$ is the static type of the
+  top level getter invocation \id{}
+  (\ref{topLevelGetterInvocation}).
+\item
+  If $D$ is a static method, library function, or local function,
+  the static type of $e$ is the function type of $D$.
```

Comment:
Here we don't address instantiated tear-offs based on context type.

Reply:
The section about generic function instantiation is written in such a way that it relies on "a compile-time error occurs except when the following step succeeds:" and then it explains how to perform generic function instantiation based on the types that we found for the target term (which is required to be a generic function type).

So it is actually consistent that we don't mention generic function instantiation right here.

I added a commentary in order to remind the reader that it may occur.